src/ssl/bio.cc

```diff
@@ -80,6 +80,12 @@ Ssl::Bio::Create(const int fd, Security::Io::Type type)
 
     if (BIO *bio = BIO_new(useMethod)) {
         BIO_int_ctrl(bio, BIO_C_SET_FD, type, fd);
+        if (enable_ktls) {
+#if OPENSSL_KTLS_SUPPORT
+            if (auto *bio_sock = BIO_new_socket(fd, BIO_NOCLOSE))
```

Comment:
Why do we need to add this BIO? In other words, why do we need _two_ custom descriptor-reading BIOs instead of one?

Reply:
I have only a little time, so short answer is:

OpenSSL's KTLS implementations are hidden as static here:
https://github.com/openssl/openssl/blob/01088602522015906654877ad2730ce805f3f925/crypto/bio/bss_sock.c#L38

- Then BIO_new_socket() (or BIO_s_socket()) is needed if squid wants to use SSL_*, is what I thought.
https://github.com/openssl/openssl/blob/01088602522015906654877ad2730ce805f3f925/crypto/bio/bss_sock.c#L58

Reply:
Thank you for providing those details. Unfortunately, they are not enough for me to connect the dots and answer the question in this change request.

Reply:
OK, I know I jumped into there.

This is related to what I prototyped first, so I would like to explain step-by-step.

Firstly, I share a thread I read as the starting point.
https://github.com/openssl/openssl/issues/14595

Here, mattcaswell mentions:

> KTLS will be used if:
>
> OpenSSL has been compiled with the "enable-ktls" option
> KTLS has not been explicitly disabled at run time via SSL_MODE_NO_KTLS_TX and SSL_MODE_NO_KTLS_RX
> You are using a suitable KTLS aware BIO (BIO_s_connect(), or BIO_s_socket())
> KTLS is compatible with the ciphersuites, TLS protocol versions and extensions that have been negotiated (this may vary from one > > > platform to another based on kernel support)

In short, using the BIO_METHOD(s) that BIO_s_connect() or BIO_s_socket() offers is the way to activate KTLS. Also, I thought BIO_s_socket() should be the appropriate one that Squid can use to intercept traffic.

Note:
 - The type BIO_METHOD (a.k.a. struct bio_method_st) is defined in OpenSSL internal header.
 - BIO_new_socket() that returns "BIO *" which is internally linked to the KTLS-aware BIO_METHOD.
 - SSL_set_fd()/SSL_set_rfd()/SSL_set_wfd() also internally generates a BIO which links to the KTLS-aware BIO_METHOD.



Based on the knowledge, I did some prototyping.

The result was:
 1. apply SSL_set_fd(fd), and not link the squid-wrapper BIO -- failed.
 2. apply SSL_set_wfd(fd), and only link the squid-wrapper BIO for receive side by SSL_set0_rbio(ssl, fd): (partially) succeeded.
 3. opposite of 2 -- failed.

Note:
 - fd is the file descriptor which had been used by squid.
 - No.2 offloads only tx side. Also, it should have other issues that I didn't realize at that time.



I started to dig into the receive side. The resulting point was the currently concerned function Ssl::ServerBio::read().

Using BIO_new_socket() and chaining the returned BIO to the squid-wrapper BIO, with some tuning of Ssl::ServerBio::read(), should be the way to go; was the conclusion of my observation.

and thereafter, probably you know.


Regarding direct answer to your comment, I do not exclude the possibility that the squid-wrapper includes everything the KTLS-aware BIO_METHOD offers, but I concern that results in Squid re-implements every related function of OpenSSL. (I may be wrong of course.)

Reply:
Your comment was very helpful, thank you! I do not yet understand whether any custom BIO can be considered a "suitable KTLS aware BIO" by OpenSSL. Do you know what specific checks/tests OpenSSL is using to decide whether the configured BIO (chain) is "suitable"?

FWIW, I understand that we can create and chain two BIOs, one custom (via BIO_new()) and one via BIO_new_socket(), like this PR is going, and that chain will be (allegedly, in your tests) considered "KTLS aware BIO" by OpenSSL.

If creating a single "suitable" custom BIO is not practical for some reason, then another alternative to consider is complete removal of Squid custom BIO: If Squid can parse the handshake by peeking at the socket (while being called from OpenSSL), then Squid ought to be able to parse the handshake by peeking at the socket without any OpenSSL calls/guidance (and then create OpenSSL connection without any custom BIO as/if needed).

Reply:
> Do you know what specific checks/tests OpenSSL is using to decide whether the configured BIO (chain) is "suitable"?

Not directly answering to your question, but this change in squid_bio_ctrl() is the one to make the entire chain of BIOs "KTLS-aware" in this context, I think.

```
#if OPENSSL_KTLS_SUPPORT
        if (BIO_next(table))
            return BIO_ctrl(BIO_next(table), cmd, arg1, arg2);
#endif
```

It passes though unknown (by squid wrapper) control messages, including BIO_CTRL_SET_KTLS etc., to the next BIO (which is KTLS-ware in this case).
https://github.com/openssl/openssl/blob/bee1e83a18d6820b4264cbefca9ff5576ff679c2/crypto/bio/bss_sock.c#L185

For the remaining part of your comment, I may post comments later if I had time..

Reply:
The comment in OpenSSL internal header below suggests BIO_CTRL_SET_KTLS is open, however, it is suspected that this is true.

https://github.com/openssl/openssl/blob/bee1e83a18d6820b4264cbefca9ff5576ff679c2/include/internal/bio.h#L40

```
/* Changes to these internal BIOs must also update include/openssl/bio.h */
# define BIO_CTRL_SET_KTLS                      72
# define BIO_CTRL_SET_KTLS_TX_SEND_CTRL_MSG     74
# define BIO_CTRL_CLEAR_KTLS_TX_CTRL_MSG        75
```

I cannot find the BIO_CTRL_SET_KTLS message defined in my local include/openssl/bio.h, and adding the case condition actually makes the compilation fail.

```
bio.cc:628:10: error: use of undeclared identifier 'BIO_CTRL_SET_KTLS'
    case BIO_CTRL_SET_KTLS:
         ^
1 error generated.
```

If the above is the truth and I'm understanding it correctly, it is almost impossible to implement a single "KTLS-aware" custom BIO outside OpenSSL unless statically linking a specific version of OpenSSL library, since the upper layer SSL_*() uses the BIO_CTRL message but the custom BIO linked to ssl instance does not understand the message in a regular manner.

  Note: KTLS-aware here means that the BIO interoperates with other KTLS related functions of OpenSSL.

So, 

> If creating a single "suitable" custom BIO is not practical for some reason

This might have been reinforced unfortunatelly.

Reply:
> then another alternative to consider is complete removal of Squid custom BIO: If Squid can parse the handshake by peeking at the socket (while being called from OpenSSL), then Squid ought to be able to parse the handshake by peeking at the socket without any OpenSSL calls/guidance (and then create OpenSSL connection without any custom BIO as/if needed).


I wonder if this difference is not important for non-blocking i/o.


```
int Ssl::Bio::write(const char *buf, int size, BIO *table)
{
    :
    BIO_clear_retry_flags(table);
    if (result **<** 0) {
        const bool ignoreError = ignoreErrno(xerrno) != 0;
        debugs(83, 5, "error: " << xerrno << " ignored: " << ignoreError);
        if (ignoreError)
            BIO_set_retry_write(table);
    }

    return result;
}
```

```
static int sock_write(BIO *b, const char *in, int inl)
{
    :
    BIO_clear_retry_flags(b);
    if (ret **<=** 0) {
        if (BIO_sock_should_retry(ret))
            BIO_set_retry_write(b);
    }
    return ret;
}
```

Reply:
Supposing the comparator difference wasn't important, the logic still has a difference in the case result:-1 and errno:EAGAIN.
At least direct return from BIO_read() breaks the function (makes it incompatible with squid logic).
It is strongly suspected that "complete removal of Squid custom BIO" is simply possible.

work:
```
int
Ssl::Bio::read(char *buf, int size, BIO *table)
{
    errno = 0;
    int result;
#if OPENSSL_KTLS_SUPPORT
    if (BIO_next(table))
        **result =** BIO_read(BIO_next(table), buf, size);
    else
#endif
```

does not work:
```
int
Ssl::Bio::read(char *buf, int size, BIO *table)
{
    errno = 0;
    int result;
#if OPENSSL_KTLS_SUPPORT
    if (BIO_next(table))
        **return** BIO_read(BIO_next(table), buf, size);
    else
#endif
```

Reply:
My comment above was not correct.
I had to set (copy) corresponding flags to "table" for this evaluation.